### PR TITLE
Netty fixes

### PR DESCRIPTION
**Pull request by rozza**
Update Netty to 4.1.5.final
Fix leak in NettyStream
Made connection string streamType the authorative setting for the streamFactoryFactory
Added deprecation warning in NettyByteBuf

JAVA-2302

### CODE REVIEW

**Comment by rozza** [comment]
Following the new [leak detection](http://netty.io/wiki/reference-counted-objects.html#troubleshooting-buffer-leaks) in netty 4.1.x the source of the leak was found.

Post applying the patch no leaks were reported with the PARANOID setting.

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java [hunk 1]
```diff
@@ -174,6 +174,7 @@ public void writeAsync(final List<ByteBuf> buffers, final AsyncCompletionHandler
             io.netty.buffer.ByteBuf byteBuf = ((NettyByteBuf) cur).asByteBuf();
             composite.addComponent(byteBuf.retain());
             composite.writerIndex(composite.writerIndex() + byteBuf.writerIndex());
+            byteBuf.release();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java]: If the retain two lines up is removed, the reference count will work out to the same as if it's retained and released

Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java]: Looking a bit more closely, I'm thinking there should be a release of 'composite' after writeAndFlush completes.

File: driver-async/src/main/com/mongodb/async/client/MongoClients.java [hunk 2]
```diff
@@ -155,6 +156,8 @@ public static MongoClient create(final ConnectionString connectionString, final
         if (connectionString.getStreamType() != null) {
             if (connectionString.getStreamType().toLowerCase().equals("netty")) {
                 builder.streamFactoryFactory(NettyStreamFactoryFactory.builder().build());
+            } else if (connectionString.getStreamType().toLowerCase().equals("nio2")) {
+                builder.streamFactoryFactory(new AsynchronousSocketChannelStreamFactoryFactory());
```
Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Can we test this in MongoClientsSpecification by temporarily setting the 'org.mongodb.async.type' system property to 'netty', and then assert that the default stream factory factory in MongoClientSettings.Builder is changed back to an instance of AsynchronousSocketChannelStreamFactoryFactory()?

File: driver-core/src/main/com/mongodb/connection/netty/NettyStream.java (hunk 1 above)
Comment (jyemin) [inline driver-core/src/main/com/mongodb/connection/netty/NettyStream.java]: Acknowledged

File: driver-async/src/main/com/mongodb/async/client/MongoClients.java (hunk 2 above)
Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Will do I triggered it by running the driver-async tests.  Jenkins runs the driver tests with netty - so it was missed.

Comment (rozza) [inline driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Done

Comment (jyemin) [inline driver-async/src/main/com/mongodb/async/client/MongoClients.java]: Nit: call the variable in the where clause 'streamFactoryFactoryClass'

**Comment by jyemin** [comment]
LGTM

**Pull request closed without merging**